# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/style/FontFaceSetImpl.cpp
# Commit: 5a39eb924188
# Full Hash: 5a39eb924188e504dad274a2148543c8d51dbae8
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2022-06-29 03:44:30
# Regressor Bug: 1771493
# File Overlap Count: 1
# Description:
#   Bug 1771493 - Part 2. Merge FontFaceSetImpl and gfxUserFontSet into a single class. r=emilio
#   
#   This patch merges FontFaceSetImpl and gfxUserFontSet into the same
#   class. This reduces the level of indirection and in theory makes it
#   easier to manage future threading issues.
# ==============================================================================

diff -r 3d30a8d83d77 -r 5a39eb924188 layout/style/FontFaceSetImpl.cpp
--- a/layout/style/FontFaceSetImpl.cpp	Tue Jun 28 21:46:59 2022 +0000
+++ b/layout/style/FontFaceSetImpl.cpp	Tue Jun 28 21:47:00 2022 +0000
@@ -127,8 +127,6 @@
   }
 
   mDocument->CSSLoader()->AddObserver(this);
-
-  mUserFontSet = new UserFontSet(this);
 }
 
 void FontFaceSetImpl::Destroy() {
@@ -152,10 +150,7 @@
   mRuleFaces.Clear();
   mNonRuleFaces.Clear();
 
-  if (mUserFontSet) {
-    mUserFontSet->mFontFaceSet = nullptr;
-    mUserFontSet = nullptr;
-  }
+  gfxUserFontSet::Destroy();
 }
 
 void FontFaceSetImpl::RemoveDOMContentLoadedListener() {
@@ -223,7 +218,7 @@
 
     const auto& name = fontFamilyName.AsFamilyName();
     RefPtr<gfxFontFamily> family =
-        mUserFontSet->LookupFamily(nsAtomCString(name.name.AsAtom()));
+        LookupFamily(nsAtomCString(name.name.AsAtom()));
 
     if (!family) {
       continue;
@@ -475,8 +470,6 @@
 
 bool FontFaceSetImpl::UpdateRules(
     const nsTArray<nsFontFaceRuleContainer>& aRules) {
-  MOZ_ASSERT(mUserFontSet);
-
   // If there was a change to the mNonRuleFaces array, then there could
   // have been a modification to the user font set.
   bool modified = mNonRuleFacesDirty;
@@ -505,7 +498,7 @@
   // the same font entries as before. (The order can affect font selection
   // where multiple faces match the requested style, perhaps with overlapping
   // unicode-range coverage.)
-  for (const auto& fontFamily : mUserFontSet->mFontFamilies.Values()) {
+  for (const auto& fontFamily : mFontFamilies.Values()) {
     fontFamily->DetachFontEntries();
   }
 
@@ -541,7 +534,7 @@
 
   // Remove any residual families that have no font entries (i.e., they were
   // not defined at all by the updated set of @font-face rules).
-  for (auto it = mUserFontSet->mFontFamilies.Iter(); !it.Done(); it.Next()) {
+  for (auto it = mFontFamilies.Iter(); !it.Done(); it.Next()) {
     if (!it.Data()->FontListLength()) {
       it.Remove();
     }
@@ -586,25 +579,19 @@
   }
 
   // if local rules needed to be rebuilt, they have been rebuilt at this point
-  if (mUserFontSet->mRebuildLocalRules) {
-    mUserFontSet->mLocalRulesUsed = false;
-    mUserFontSet->mRebuildLocalRules = false;
+  if (mRebuildLocalRules) {
+    mLocalRulesUsed = false;
+    mRebuildLocalRules = false;
   }
 
   if (LOG_ENABLED() && !mRuleFaces.IsEmpty()) {
-    LOG(("userfonts (%p) userfont rules update (%s) rule count: %d",
-         mUserFontSet.get(), (modified ? "modified" : "not modified"),
-         (int)(mRuleFaces.Length())));
+    LOG(("userfonts (%p) userfont rules update (%s) rule count: %d", this,
+         (modified ? "modified" : "not modified"), (int)(mRuleFaces.Length())));
   }
 
   return modified;
 }
 
-void FontFaceSetImpl::IncrementGeneration(bool aIsRebuild) {
-  MOZ_ASSERT(mUserFontSet);
-  mUserFontSet->IncrementGeneration(aIsRebuild);
-}
-
 void FontFaceSetImpl::InsertNonRuleFontFace(FontFaceImpl* aFontFace,
                                             bool& aFontSetModified) {
   nsAtom* fontFamily = aFontFace->GetFamilyName();
@@ -618,8 +605,7 @@
 
   // Just create a new font entry if we haven't got one already.
   if (!aFontFace->GetUserFontEntry()) {
-    // XXX Should we be checking mUserFontSet->mLocalRulesUsed like
-    // InsertRuleFontFace does?
+    // XXX Should we be checking mLocalRulesUsed like InsertRuleFontFace does?
     RefPtr<gfxUserFontEntry> entry = FindOrCreateUserFontEntryFromFontFace(
         family, aFontFace, StyleOrigin::Author);
     if (!entry) {
@@ -629,7 +615,7 @@
   }
 
   aFontSetModified = true;
-  mUserFontSet->AddUserFontEntry(family, aFontFace->GetUserFontEntry());
+  AddUserFontEntry(family, aFontFace->GetUserFontEntry());
 }
 
 void FontFaceSetImpl::InsertRuleFontFace(FontFaceImpl* aFontFace,
@@ -658,7 +644,7 @@
     if (rec.mFontFace == aFontFace && rec.mOrigin == Some(aSheetType)) {
       // if local rules were used, don't use the old font entry
       // for rules containing src local usage
-      if (mUserFontSet->mLocalRulesUsed && mUserFontSet->mRebuildLocalRules) {
+      if (mLocalRulesUsed && mRebuildLocalRules) {
         if (aFontFace->HasLocalSrc()) {
           // Remove the old record, but wait to see if we successfully create a
           // new user font entry below.
@@ -671,7 +657,7 @@
       gfxUserFontEntry* entry = rec.mFontFace->GetUserFontEntry();
       MOZ_ASSERT(entry, "FontFace should have a gfxUserFontEntry by now");
 
-      mUserFontSet->AddUserFontEntry(family, entry);
+      AddUserFontEntry(family, entry);
 
       MOZ_ASSERT(!HasRuleFontFace(rec.mFontFace),
                  "FontFace should not occur in mRuleFaces twice");
@@ -733,7 +719,7 @@
   // on the family, gfxUserFontFamily::AddFontEntry(), which AddUserFontEntry
   // calls, will automatically remove the earlier occurrence of the same
   // userfont entry.
-  mUserFontSet->AddUserFontEntry(family, entry);
+  AddUserFontEntry(family, entry);
 }
 
 /* static */
@@ -875,8 +861,7 @@
     // family.
     if (!existingEntry->mFamilyName.IsEmpty() &&
         existingEntry->mFamilyName != aFamilyName) {
-      gfxUserFontFamily* family =
-          set->GetUserFontSet()->LookupFamily(existingEntry->mFamilyName);
+      gfxUserFontFamily* family = set->LookupFamily(existingEntry->mFamilyName);
       if (family) {
         family->RemoveFontEntry(existingEntry);
       }
@@ -1002,7 +987,7 @@
     return nullptr;
   }
 
-  RefPtr<gfxUserFontEntry> entry = set->mUserFontSet->FindOrCreateUserFontEntry(
+  RefPtr<gfxUserFontEntry> entry = set->FindOrCreateUserFontEntry(
       aFamilyName, srcArray, weight, stretch, italicStyle, featureSettings,
       variationSettings, languageOverride, unicodeRanges, fontDisplay,
       rangeFlags, ascentOverride, descentOverride, lineGapOverride, sizeAdjust);
@@ -1079,7 +1064,7 @@
   message.AppendLiteral(" source: ");
   message.Append(fontURI);
 
-  LOG(("userfonts (%p) %s", mUserFontSet.get(), message.get()));
+  LOG(("userfonts (%p) %s", this, message.get()));
 
   // try to give the user an indication of where the rule came from
   RawServoFontFaceRule* rule = FindRuleForUserFontEntry(aUserFontEntry);
@@ -1127,12 +1112,8 @@
 }
 
 void FontFaceSetImpl::CacheFontLoadability() {
-  if (!mUserFontSet) {
-    return;
-  }
-
   // TODO(emilio): We could do it a bit more incrementally maybe?
-  for (const auto& fontFamily : mUserFontSet->mFontFamilies.Values()) {
+  for (const auto& fontFamily : mFontFamilies.Values()) {
     fontFamily->ReadLock();
     for (const gfxFontEntry* entry : fontFamily->GetFontList()) {
       if (!entry->mIsUserFontContainer) {
@@ -1164,17 +1145,13 @@
 
   MOZ_ASSERT(NS_IsMainThread());
 
-  if (!mUserFontSet) {
-    return false;
-  }
-
   if (aSrc.mUseOriginPrincipal) {
     return true;
   }
 
-  gfxFontSrcPrincipal* gfxPrincipal = aSrc.mURI->InheritsSecurityContext()
-                                          ? nullptr
-                                          : aSrc.LoadPrincipal(*mUserFontSet);
+  RefPtr<gfxFontSrcPrincipal> gfxPrincipal =
+      aSrc.mURI->InheritsSecurityContext() ? nullptr
+                                           : aSrc.LoadPrincipal(*this);
 
   nsIPrincipal* principal =
       gfxPrincipal ? gfxPrincipal->NodePrincipal() : nullptr;
@@ -1194,21 +1171,6 @@
   return NS_SUCCEEDED(rv) && NS_CP_ACCEPTED(shouldLoad);
 }
 
-void FontFaceSetImpl::DispatchFontLoadViolations(
-    nsTArray<nsCOMPtr<nsIRunnable>>& aViolations) {
-  if (XRE_IsContentProcess()) {
-    nsCOMPtr<nsIEventTarget> eventTarget =
-        mDocument->EventTargetFor(TaskCategory::Other);
-    for (nsIRunnable* runnable : aViolations) {
-      eventTarget->Dispatch(do_AddRef(runnable), NS_DISPATCH_NORMAL);
-    }
-  } else {
-    for (nsIRunnable* runnable : aViolations) {
-      NS_DispatchToMainThread(do_AddRef(runnable));
-    }
-  }
-}
-
 nsresult FontFaceSetImpl::SyncLoadFontData(gfxUserFontEntry* aFontToLoad,
                                            const gfxFontFaceSrc* aFontFaceSrc,
                                            uint8_t*& aBuffer,
@@ -1488,7 +1450,7 @@
   }
 }
 
-nsPresContext* FontFaceSetImpl::GetPresContext() {
+nsPresContext* FontFaceSetImpl::GetPresContext() const {
   if (!mDocument) {
     return nullptr;
   }
@@ -1501,48 +1463,19 @@
   mStandardFontLoadPrincipal = new gfxFontSrcPrincipal(
       mDocument->NodePrincipal(), mDocument->PartitionedPrincipal());
   mAllowedFontLoads.Clear();
-  if (mUserFontSet) {
-    mUserFontSet->IncrementGeneration(false);
-  }
-}
-
-// -- FontFaceSetImpl::UserFontSet
-// ------------------------------------------------
-
-/* virtual */
-bool FontFaceSetImpl::UserFontSet::IsFontLoadAllowed(
-    const gfxFontFaceSrc& aSrc) {
-  return mFontFaceSet && mFontFaceSet->IsFontLoadAllowed(aSrc);
+  IncrementGeneration(false);
 }
 
-/* virtual */
-void FontFaceSetImpl::UserFontSet::DispatchFontLoadViolations(
-    nsTArray<nsCOMPtr<nsIRunnable>>& aViolations) {
-  if (mFontFaceSet) {
-    mFontFaceSet->DispatchFontLoadViolations(aViolations);
-  }
-}
+// -- gfxUserFontSet
+// ------------------------------------------------
 
-/* virtual */
-nsresult FontFaceSetImpl::UserFontSet::StartLoad(
-    gfxUserFontEntry* aUserFontEntry, uint32_t aSrcIndex) {
-  if (!mFontFaceSet) {
-    return NS_ERROR_FAILURE;
-  }
-  return mFontFaceSet->StartLoad(aUserFontEntry, aSrcIndex);
-}
-
-void FontFaceSetImpl::UserFontSet::RecordFontLoadDone(uint32_t aFontSize,
-                                                      TimeStamp aDoneTime) {
+void FontFaceSetImpl::RecordFontLoadDone(uint32_t aFontSize,
+                                         TimeStamp aDoneTime) {
   mDownloadCount++;
   mDownloadSize += aFontSize;
   Telemetry::Accumulate(Telemetry::WEBFONT_SIZE, aFontSize / 1024);
 
-  if (!mFontFaceSet) {
-    return;
-  }
-
-  TimeStamp navStart = mFontFaceSet->GetNavigationStartTimeStamp();
+  TimeStamp navStart = GetNavigationStartTimeStamp();
   TimeStamp zero;
   if (navStart != zero) {
     Telemetry::AccumulateTimeDelta(Telemetry::WEBFONT_DOWNLOAD_TIME_AFTER_START,
@@ -1550,44 +1483,9 @@
   }
 }
 
-/* virtual */
-nsresult FontFaceSetImpl::UserFontSet::LogMessage(
-    gfxUserFontEntry* aUserFontEntry, uint32_t aSrcIndex, const char* aMessage,
-    uint32_t aFlags, nsresult aStatus) {
-  if (!mFontFaceSet) {
-    return NS_ERROR_FAILURE;
-  }
-  return mFontFaceSet->LogMessage(aUserFontEntry, aSrcIndex, aMessage, aFlags,
-                                  aStatus);
-}
+void FontFaceSetImpl::DoRebuildUserFontSet() { MarkUserFontSetDirty(); }
 
-/* virtual */
-nsresult FontFaceSetImpl::UserFontSet::SyncLoadFontData(
-    gfxUserFontEntry* aFontToLoad, const gfxFontFaceSrc* aFontFaceSrc,
-    uint8_t*& aBuffer, uint32_t& aBufferLength) {
-  if (!mFontFaceSet) {
-    return NS_ERROR_FAILURE;
-  }
-  return mFontFaceSet->SyncLoadFontData(aFontToLoad, aFontFaceSrc, aBuffer,
-                                        aBufferLength);
-}
-
-/* virtual */
-bool FontFaceSetImpl::UserFontSet::GetPrivateBrowsing() {
-  return mFontFaceSet && mFontFaceSet->mPrivateBrowsing;
-}
-
-/* virtual */
-void FontFaceSetImpl::UserFontSet::DoRebuildUserFontSet() {
-  if (!mFontFaceSet) {
-    return;
-  }
-  mFontFaceSet->MarkUserFontSetDirty();
-}
-
-/* virtual */
-already_AddRefed<gfxUserFontEntry>
-FontFaceSetImpl::UserFontSet::CreateUserFontEntry(
+already_AddRefed<gfxUserFontEntry> FontFaceSetImpl::CreateUserFontEntry(
     const nsTArray<gfxFontFaceSrc>& aFontFaceSrcList, WeightRange aWeight,
     StretchRange aStretch, SlantStyleRange aStyle,
     const nsTArray<gfxFontFeature>& aFeatureSettings,